Kernel/Net/TCPSocket.cpp

```diff
@@ -664,6 +664,30 @@ ErrorOr<void> TCPSocket::close()
 {
     MutexLocker locker(mutex());
     auto result = IPv4Socket::close();
+    if (m_linger.l_onoff != 0) {
+        auto has_unacked_data = m_unacked_packets.with_shared([&](auto const& packets) { return packets.size > 0; });
+        if (m_linger.l_linger > 0) {
+            dbgln_if(TCP_SOCKET_DEBUG, "SO_LINGER enabled. Lingering for {} seconds", m_linger.l_linger);
+            auto deadline = TimeManagement::the().monotonic_time() + Duration::from_seconds(m_linger.l_linger);
+            do {
+                if (!has_unacked_data)
+                    break;
+                (void)Thread::current()->sleep(Duration::from_milliseconds(1));
+            } while (TimeManagement::the().monotonic_time() < deadline);
+
+            if (has_unacked_data) {
```

Comment:
you never re-check the unacked data, meaning this variable will likely not correctly reflect the current state of things at this point (likely, most data was acked by now, but that’s not checked).